### PR TITLE
V3.x simplify conversion factors

### PR DESCRIPTION
**Pull request by nholthaus**
This PR removes the symbols naming conversion factors for each unit container type.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,86 +1,86 @@
...
-    DESTINATION share/units/cmake
-)
-
-install(DIRECTORY include/
-    DESTINATION include
-)
+cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
+
+PROJECT(units LANGUAGES CXX)
+set(units_VERSION 3.0.0)
```
Comment (johelegp) [inline CMakeLists.txt:4]:
```suggestion
project(units VERSION 3.0.0 LANGUAGES CXX)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,86 +1,86 @@
...
+
+# cmake options
+OPTION(BUILD_TESTS "Build unit tests" ON)
```
Comment (johelegp) [inline CMakeLists.txt:7]:
```suggestion
OPTION(BUILD_TESTS "Build unit tests" OFF)
```
This is more friendly to users. Developers, the CI and interested users should explicitly turn it on.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,86 +1,86 @@
...
+OPTION(BUILD_DOCS "Build the documentation" OFF)
+OPTION(DISABLE_IOSTREAM "Disables <iostream> (cout) support for embedded applications" OFF)
```
Comment (johelegp) [inline CMakeLists.txt:9]: Consider prefixing `UNITS_` to the options to avoid clashes with other libraries.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,86 +1,86 @@
...
+
+SET(CMAKE_CXX_STANDARD_REQUIRED TRUE)
```
Comment (johelegp) [inline CMakeLists.txt:11]: Similar to above, the library doesn't strictly require this. Also see #245.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,86 +1,86 @@
...
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+	add_compile_options(-Wall -Wextra -pedantic)
+elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+	# enable parallel build
+	set( ENV{CL} /MP )
+#	target_compile_definitions(${PROJECT_NAME} INTERFACE "/permissive-")
+elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+	# Increase warning levels
+	add_compile_options(-Wall -Wextra -pedantic)
+endif()
```
Comment (johelegp) [inline CMakeLists.txt:35]: This, too, belong in the CI and developer's instructions. Then you can use `-Werror` without remorse.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,86 +1,86 @@
...
+
+# Remove IOStream from the library (useful for embdedded development)
+if(DISABLE_IOSTREAM)
+	target_compile_definitions(${PROJECT_NAME} INTERFACE -DUNIT_LIB_DISABLE_IOSTREAM)
+endif(DISABLE_IOSTREAM)
+
+# unit tests
+if(BUILD_TESTS)
+	enable_testing()
```
Comment (johelegp) [inline CMakeLists.txt:44]:
```suggestion
include(CTest)
if(BUILD_TESTING AND BUILD_TESTS)
```
See https://cmake.org/cmake/help/latest/module/CTest.html.

File: .travis.yml [hunk 7]
```diff
@@ -44,9 +44,9 @@ install:
 script:
   - mkdir build && cd build
   - cmake --version
-  - cmake ..
-  - cmake --build . --config Release
-  - cmake --build . --config Debug
+  - cmake _DUNITS_BUILD_TESTS=ON ..
```
Comment (johelegp) [inline .travis.yml:47]:
```suggestion
  - cmake -DUNITS_BUILD_TESTS=ON ..
```

**Comment by johelegp** [comment]
The change is really hard to see. Can you rebase it on `v3.x`?

**Comment by nholthaus** [comment]
rebased to v3.x. coveralls isn't working (maybe hasn't been for a while?) but I don't want that to hold up the rename, and I think it would be better to solve it in another PR.

**Comment by johelegp** [comment]
What rename? The units were renamed and this removes the conversion factors. I can tell you did that from the tests, but I can't see how you did it because the diff for `core.h` is messed up.

**Comment by nholthaus** [comment]
Whitespace issue, I was hoping the rebase would help. I'll take care of that.

You can view the differ [without the whitespace change](https://github.com/nholthaus/units/pull/256/files?w=1)

File: include/units/core.h [hunk 8]
```diff
@@ -1,3922 +1,3892 @@
...
+//	MACROS
+//------------------------------
+
+/**
+ * @def			UNIT_ADD_UNIT_DEFINITION(namespaceName,namePlural)
+ * @brief		Macro for generating the boiler-plate code for the unit type definition.
+ * @details		The macro generates the definition of the unit container types, e.g. `meter`
+ * @param		namespaceName namespace in which the new units will be encapsulated.
+ * @param		namePlural - plural version of the unit name, e.g. 'meters'
+ */
```
Comment (johelegp) [inline include/units/core.h:130]:
```suggestion
 * @param		__VA_ARGS__ the conversion factor of the unit, e.g. `meters<>`
 */
```

File: include/units/core.h [hunk 9]
```diff
@@ -1,3922 +1,3892 @@
...
+ * @details		The macro generates the definition of the scaled unit templates as a strong type template alias,
+ *				e.g. `meters`
+ * @param		unitName unit name, e.g. 'meters'
+ * @param		scale the non linear scale template argument of the unit's base
+ * @param		definition - the variadic parameter is used for the definition of the unit
+ *				(e.g. `conversion_factor<std::ratio<1>, units::dimension::length>`)
+ * @note		a variadic template is used for the definition to allow templates with
+ *				commas to be easily expanded. All the variadic 'arguments' should together
+ *				comprise the unit definition.
+ */
```
Comment (johelegp) [inline include/units/core.h:150]: Like above.

File: include/units/core.h [hunk 10]
```diff
@@ -1,3922 +1,3892 @@
...
+ * @param		namespaceName namespace in which the new units will be encapsulated. All literal values
+ *				are placed in the `units::literals` namespace.
+ * @param		namePlural - plural version of the unit name, e.g. 'meters'
+ * @param		abbreviation - abbreviated unit name, e.g. 'm'
+ * @param		definition - the variadic parameter is used for the definition of the unit
+ *				(e.g. `conversion_factor<std::ratio<1>, units::dimension::length>`)
+ * @note		a variadic template is used for the definition to allow templates with
+ *				commas to be easily expanded. All the variadic 'arguments' should together
+ *				comprise the unit definition.
+ */
```
Comment (johelegp) [inline include/units/core.h:249]: Like above.

**Comment by johelegp** [comment]
How are the error messages after this? `meters` is actually `unit<...>`, so I suspect they'll revert the compiler to being a novelist.

**Comment by nholthaus** [comment]
I need to check, when I started we still had strong units.

That said, conversion names didn't appear in the error messages as often a
you might think, hence why we used to see all the old SI dimension ratios.

**Comment by johelegp** [comment]
> That said, conversion names didn't appear in the error messages as often a
you might think, hence why we used to see all the old SI dimension ratios.

Seems like I didn't implement the follow up to identify strong units within ephemeral units. I suspect it's not easy to do, anyways.

**Comment by nholthaus** [comment]
The latest push keeps the readability improvements in the definition, while retaining the improved error messages. Clang is a bit inconsistent about when it uses named strong types, but overall still improved.

We weren't able to get rid of any symbols, but at least they've become a pure implementation detail. 

I think this is ready to go.

**Comment by nholthaus** [comment]
resolves #209

**Pull request merged**